### PR TITLE
Better awful.rules doc

### PR DESCRIPTION
**Pull request by Elv13**
What's done:

 * Move the documentation parser into a module
     * Add C documentation support to the parser
     * Make the parser more flexible
 * Update the CSS to render the parser output in a similar way to the other components
 * Add a table of all client properties to `awful.rules` and `awful.spawn`
 * Refactor `awful.rules` documentation
 * Refactor `awful.spawn` documentation

This pull request is mostly moving code around and adding boilerplate `ldoc` packaging for the rules own properties.

~~There is still somehow a target dependency issue where the new rules table only exists the second time the doc is built. It might just be late, but I can't see it right now.(edit: I will need to move `configure_file` into a target, but it wont be easy given it uses many variables)

Also the links are often broken in the table because `awful.client` redirects to `client` and the parser is too dumb to notice. I think it's worth merging even with that bug, but the dependency issue needs to be solved.~~ 

Finally, if anyone disagree with the new documentation, just edit the pull request instead of making comments.

### CODE REVIEW

**Comment by Elv13** [comment]
Updated/Finished

The hack to keep the client property table up-to-date isn't pretty, but changing the build system to get rid of `configure_file` and really, really, do nothing during the configuration phase is too large for this PR. So for now it just make sure to get the file in the commit.

Beside that I updated the CSS style to make the tables blend better, it follows @actionless initial work on this by adding an header style and padding that match GitHub issues tables. It also add `<code>` and `<kdb>` style to also match GitHub ones.

File: lib/awful/client.lua [hunk 1]
```diff
@@ -59,6 +59,93 @@ client.property = {}
...
+--
+-- The `request::activate` context for this call is `rules`.
+--
+-- @clientruleproperty focus
+-- @param[opt=false] boolean
+
+--- Should this client have a titlebar by default.
+-- @clientruleproperty titlebars_enabled
+-- @param[opt=false] boolean
+-- @see awful.titlebar
```
Comment (psychon) [inline lib/awful/client.lua]: Arguably, all the `@param boolean` above are wrong, since the code also supports calling functions that then return a boolean (IIRC).

File: awesomeConfig.cmake [hunk 2]
```diff
@@ -354,13 +354,24 @@ add_custom_target(lgi-check-run ALL
 
 # {{{ Generate some aggregated documentation from lua script
 
-file(MAKE_DIRECTORY ${BUILD_DIR}/script_files/)
+# file(MAKE_DIRECTORY ${BUILD_DIR}/script_files/)
+# file(MAKE_DIRECTORY ${BUILD_DIR}/docs/common/)
+# file(COPY ${SOURCE_DIR}/docs/_parser.lua DESTINATION ${BUILD_DIR}/docs/)
```
Comment (psychon) [inline awesomeConfig.cmake]: Why these new comments?

File: awesomeConfig.cmake [hunk 3]
```diff
@@ -354,13 +354,24 @@ add_custom_target(lgi-check-run ALL
...
+
+add_custom_target(setup_directories DEPENDS lgi-check-run)
+
+add_custom_command(TARGET setup_directories
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${BUILD_DIR}/script_files/
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${BUILD_DIR}/docs/common/
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${BUILD_DIR}/doc/images/
+        COMMAND ${CMAKE_COMMAND} -E copy ${SOURCE_DIR}/docs/_parser.lua ${BUILD_DIR}/docs/
```
Comment (psychon) [inline awesomeConfig.cmake]: Why these new statements? I cannot find anything creating these directories previously.

File: docs/_parser.lua [hunk 4]
```diff
@@ -0,0 +1,196 @@
+local gio = require("lgi").Gio
```
Comment (psychon) [inline docs/_parser.lua]: Could you move the addition of this file and the new `COPY` for it (why is that needed?) to the next commit? That way, it doesn't look like just new code and one can actually see which part was just moved and which is actually new.

File: CMakeLists.txt [hunk 5]
```diff
@@ -289,7 +289,11 @@ if(GENERATE_DOC)
     file(COPY ${SOURCE_DIR}/docs/ldoc.css DESTINATION ${BUILD_DIR}/docs)
 
     add_custom_target(ldoc ALL
-        DEPENDS ${BUILD_DIR}/doc/index.html ${BUILD_DIR}/docs/ldoc.css)
+        DEPENDS
+            ${BUILD_DIR}/doc/index.html
+            ${BUILD_DIR}/docs/ldoc.css
+            generate_awesomerc
+    )
```
Comment (psychon) [inline CMakeLists.txt]: I'm not sure if this change should be in this commit. It doesn't have anything to do with the index, I think. Perhaps the previous commit (which already has some unexplained dependency additions/fixes anyway).

File: awesomeConfig.cmake [hunk 6]
```diff
@@ -371,7 +371,26 @@ add_custom_command(
...
+
+add_custom_command(
+        OUTPUT ${BUILD_DIR}/docs/common/rules_index.ldoc
+        COMMAND lua ${SOURCE_DIR}/docs/build_rules_index.lua
+            ${BUILD_DIR}/docs/common/rules_index.ldoc
+
+        # Cheap trick until the ldoc `configure_file` is ported to be a build
+        # step rather than part of cmake.
+        COMMAND ${CMAKE_COMMAND} -E copy ${BUILD_DIR}/docs/common/rules_index.ldoc
+            ${SOURCE_DIR}/docs/common/rules_index.ldoc
```
Comment (psychon) [inline awesomeConfig.cmake]: Uhm... Does this copy something into the *source* directory? Why?!?
Isn't the build supposed not to touch the source directory in any way? Autoconf even checks for that during `make distcheck` by testing an out-of-tree build against read-only sources.

File: docs/06-appearance.md.lua [hunk 7]
```diff
@@ -229,8 +70,9 @@ f:write[[
 
 Beautiful is where Awesome theme variables are stored.
 
+<br \><br \>
 ]]
-f:write(create_table(beautiful_vars, {"link", "desc"}))
+f:write(parser.create_table(beautiful_vars, {"link", "desc"}))
```
Comment (psychon) [inline docs/06-appearance.md.lua]: Looks like this only changes the code to use the new `parser` (so should be done in the commit that adds `parser`, I guess)... and then, some new `<br \><br \>` appear that seem totally unrelated. Why?

File: docs/06-appearance.md.lua [hunk 8]
```diff
@@ -229,8 +70,9 @@ f:write[[
 
 Beautiful is where Awesome theme variables are stored.
 
+<br \><br \>
```
Comment (psychon) [inline docs/06-appearance.md.lua]: Shouldn't it be `<br />` instead of `<br \>`?

File: docs/common/rules_index.ldoc [hunk 9]
```diff
@@ -0,0 +1,72 @@
+--<table class='widget_list' border=1>
```
Comment (psychon) [inline docs/common/rules_index.ldoc]: Why is this file added to the source if it is generated during the build?

File: docs/load_ldoc.cmake [hunk 10]
```diff
@@ -1,23 +1,23 @@
 # To avoid copy pasting, some documentation is stored in reusable files
 set(SHAPE_FILE "${SOURCE_DIR}/docs/common/${SHAPE_NAME}.lua")
 
-set(path "${SOURCE_DIR}/docs/common/")
+foreach(path ${BUILD_DIR}/docs/common/;${SOURCE_DIR}/docs/common/)
```
Comment (psychon) [inline docs/load_ldoc.cmake]: This is just for `rules_index.ldoc`, right? If so, when that file is generated in the build dir, then this piece of code is not needed.

(Also, I just remember that `rules_index.ldoc` is copied around... doesn't that mean that ldoc ends up parsing both the in-source and in-build copy of the file?)

File: lib/awful/rules.lua [hunk 11]
```diff
@@ -8,17 +8,12 @@
...
--- * placement
--- * honor_padding
--- * honor_workarea
--- * tag
--- * new_tag
--- * switch_to_tags (also called switchtotag)
--- * focus
--- * titlebars_enabled
--- * callback
+-- This table contains the client properties that are valid when used in rules.
```
Comment (psychon) [inline lib/awful/rules.lua]: for use in rules?

But the whole sentence strikes me as a bit odd. You can add anything to rules and it is just set as a property on the client...

File: lib/awful/spawn.lua [hunk 12]
```diff
@@ -45,45 +48,6 @@
...
---
--- This can also be used from the command line:
---
---    awesome-client 'awful=require("awful");
---      awful.spawn("urxvt -e maxima -name CALCULATOR", {
---        floating  = true,
---        tag       = mouse.screen.selected_tag,
---        placement = awful.placement.bottom_right,
---      })'
---
```
Comment (psychon) [inline lib/awful/spawn.lua]: This change seems unrelated to this commit...

File: lib/awful/spawn.lua [hunk 13]
```diff
@@ -48,6 +48,57 @@
...
+-- command. It is strongly recommanded not to use `io.popen` is explained in the
+-- "Getting a command's output" section. Asynchronous execution is at first a
+-- bit tricky to understand if you never used that before. The example below
+-- should demonstrate how it works.
+--
+-- If we do (but *really*, don't do that):
+--
+--     -- **NEVER, EVER, DO THIS**: your computer will freeze
+--     os.execute("sleep 1; echo foo > /tmp/foo.txt")
+--     mylabel.text = io.popen("cat /tmp/foo.txt")
```
Comment (psychon) [inline lib/awful/spawn.lua]: `popen` does not work like this, does it? It returns a file handle that you can then read from.

File: lib/awful/spawn.lua [hunk 14]
```diff
@@ -48,6 +48,57 @@
...
+--
+-- The label will display `foo`. But If we do:
+--
+--     -- Don't do this, it wont work.
+--     -- Assumes /tmp/foo.txt does not exist
+--     awful.spawn.with_shell("sleep 1; echo foo > /tmp/foo.txt")
+--     mylabel.text = io.popen("cat /tmp/foo.txt")
```
Comment (psychon) [inline lib/awful/spawn.lua]: Same

File: lib/awful/spawn.lua [hunk 15]
```diff
@@ -48,6 +48,57 @@
...
+--
+-- Then the label will be **empty**. `awful.spawn` and `awful.spawn.with_shell`
+-- will **not** block and thus the `io.popen` will be executed before
+-- `sleep 1` finishes. This is why we have async functions to execute shell
+-- commands. There is many variants with difference characteristics and
```
Comment (psychon) [inline lib/awful/spawn.lua]: There *are*

File: lib/awful/client.lua (hunk 1 above)
Comment (Elv13) [reply lib/awful/client.lua]: Yeah, but that's true for all rules. Treating them differently for client properties make little sense here. I don't think it is worth documenting this in a different way here.

File: awesomeConfig.cmake (hunk 2 above)
Comment (Elv13) [reply awesomeConfig.cmake]: done

File: awesomeConfig.cmake (hunk 3 above)
Comment (Elv13) [reply awesomeConfig.cmake]: One of them is created in `tests/CMakeLists.txt`. The others by `configure_file`. It fixes #2342. I will add a note in the commit message

File: docs/_parser.lua (hunk 4 above)
Comment (Elv13) [reply docs/_parser.lua]: > why is that needed?

To avoid messing with LUA_PATH. I can use either solution, I don't mind.

File: CMakeLists.txt (hunk 5 above)
Comment (Elv13) [reply CMakeLists.txt]: done

File: awesomeConfig.cmake (hunk 6 above)
Comment (Elv13) [reply awesomeConfig.cmake]: Not pretty, I will agree on that. The problem is that `configure_file` depends on it and fixing that will be non-trivial. tl;dr, we will need a `configure_file` equivalent implemented in Lua. It is worth doing and generally would make sense anyway. However I chose not to do it in this pull request as I want v4.3 released and this is part 2 of 4 of the "improve spawn" serie.

File: docs/06-appearance.md.lua (hunk 7 above)
Comment (Elv13) [reply docs/06-appearance.md.lua]: > appear that seem totally unrelated. Why?

It was already there https://github.com/awesomeWM/awesome/blob/master/docs/06-appearance.md.lua#L166

File: docs/06-appearance.md.lua (hunk 8 above)
Comment (Elv13) [reply docs/06-appearance.md.lua]: done

File: docs/common/rules_index.ldoc (hunk 9 above)
Comment (Elv13) [reply docs/common/rules_index.ldoc]: dependency loop, see above

File: docs/load_ldoc.cmake (hunk 10 above)
Comment (Elv13) [reply docs/load_ldoc.cmake]: see previous comment

File: lib/awful/rules.lua (hunk 11 above)
Comment (Elv13) [reply lib/awful/rules.lua]: done

File: lib/awful/spawn.lua (hunk 12 above)
Comment (Elv13) [reply lib/awful/spawn.lua]: done

File: lib/awful/spawn.lua (hunk 13 above)
Comment (Elv13) [reply lib/awful/spawn.lua]: done

File: lib/awful/spawn.lua (hunk 14 above)
Comment (Elv13) [reply lib/awful/spawn.lua]: done

**Comment by Elv13** [comment]
Updated

**Comment by psychon** [comment]
Shall I take another look or shall we do this the easy way instead?

**Comment by Elv13** [comment]
> Shall I take another look or shall we do this the easy way instead?

I guess the "it modifies the source directory" is still a rather controversial issue. If you are fine with my "I am lazy" explanation, press the merge button.

**Comment by Elv13** [comment]
I guess that's a yes. If nobody else cares to review, I will merge that tomorrow so I can push the next PR of this serie (and last for v4.3, the last one refactor the `.desktop` support so you can use them in spawn, I am not merging that for 4.3)
